File: QytetetJava/modeloqytetet/Casilla.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

/**
 *
 * @author csp98
 */
public abstract class Casilla {

    protected int numeroCasilla;
    protected int coste;
    protected TipoCasilla tipo;

    // private int numHoteles = 0;
    // private int numCasas = 0;
    // private TituloPropiedad titulo;
    public Casilla(int num, int c, TipoCasilla type) { //Cárcel, salida, etc. No se compran
        this.numeroCasilla = num;
        this.coste = c;
        this.tipo = type;
        // this.titulo = null;
    }

    public int getNumeroCasilla() {
        return numeroCasilla;
    }

    public int getCoste() {
        return coste;
    }

    public TipoCasilla getTipoCasilla() {
        return tipo;
    }

    abstract boolean soyEdificable();

    @Override
    public abstract String toString();

    /*
    public Casilla(int num, int c) { //Calles (se pueden comprar)
        this.numeroCasilla = num;
        this.coste = c;
        this.tipo = TipoCasilla.CALLE;
        this.titulo = null;
    }
     
    int getNumeroHoteles() {
        return numHoteles;
    }

    int getNumeroCasas() {
        return numCasas;
    }

    public TituloPropiedad getTituloPropiedad() {
        return titulo;
    }

    void setNumHoteles(int n) {
        numHoteles = n;
    }

    void setNumCasas(int n) {
        numCasas = n;
    }

    void setTitulo(TituloPropiedad titulo) {
        this.titulo = titulo;
    }

    @Override
    public String toString() {
        String pro;
        if (this.tipo == TipoCasilla.CALLE) {
            pro = "\nTítulo de propiedad:\n" + this.titulo.toString();
        } else {
            pro = ".\n";
        }

        return "\nNumero de casilla: " + this.numeroCasilla + ".\nCoste de compra: " + this.coste
                + ".\nNúmero de casas construidas: " + this.numCasas
                + ".\nNúmero de hoteles construidos: " + this.numHoteles
                + ".\nTipo de casilla: " + this.tipo + pro;

    }

    TituloPropiedad asignarPropietario(Jugador jugador) {
        titulo.setPropietario(jugador);
        return titulo;
    }

    int calcularValorHipoteca() {
        int hipotecaBase;
        hipotecaBase = titulo.getHipotecaBase();
        int cantidadRecibida = hipotecaBase + (int) (numCasas * 0.55 * hipotecaBase + numHoteles * hipotecaBase);
        return cantidadRecibida;
    }

    int cancelarHipoteca() {
        int aPagar;
        titulo.setHipotecada(false);
        aPagar = getCosteHipoteca();
        return aPagar;
    }

    int edificarCasa() {
        int costeEdificarCasa;
        setNumCasas(numCasas + 1);
        costeEdificarCasa = titulo.getPrecioEdificar();
        return costeEdificarCasa;

    }

    int edificarHotel() {
        int costeEdificarHotel;
        setNumHoteles(numHoteles + 1);
        costeEdificarHotel = titulo.getPrecioEdificar();
        return costeEdificarHotel;
    }

    public int cobrarAlquiler() {
        int costeAlquiler;
        int costeAlquilerBase = titulo.getAlquilerBase();
        costeAlquiler = costeAlquilerBase + (int) (numCasas * 0.5 + numHoteles * 2);
        titulo.cobrarAlquiler(costeAlquiler);
        return costeAlquiler;
    }

    boolean estaHipotecada() {
        return titulo.getHipotecada();
    }

    int getCosteHipoteca() {
        int coste;
        coste = (int) (calcularValorHipoteca() * 1.10);  //Es un 10% más
        return coste;
    }

    int hipotecar() {
        int cantidadRecibida;
        titulo.setHipotecada(true);
        cantidadRecibida = calcularValorHipoteca();
        return cantidadRecibida;
    }

    boolean propietarioEncarcelado() {
        return titulo.propietarioEncarcelado();
    }

    public boolean sePuedeEdificarCasa(int factorEspeculador) {
        return numCasas < 4 * factorEspeculador;
    }

    public boolean sePuedeEdificarHotel(int factorEspeculador) {
        return numHoteles < 4 * factorEspeculador && numCasas == 4 * factorEspeculador;
    }

    boolean tengoPropietario() {
        return titulo.tengoPropietario();
    }

    int venderTitulo() {
        int precioVenta;
        int precioCompra;
        precioCompra = coste + (numCasas + numHoteles) * titulo.getPrecioEdificar();
        precioVenta = (int) (precioCompra + titulo.getFactorRevalorizacion() * precioCompra);
        titulo.setPropietario(null);
        setNumHoteles(0);
        setNumCasas(0);
        return precioVenta;
    }

     */
}

File: QytetetRuby/Casilla.rb
#encoding: utf-8

# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative 'TipoCasilla'
require_relative 'TituloPropiedad'
module ModeloQytetet
  class Casilla
    #attr_reader :coste, :numeroCasilla , :tipo
    #attr_accessor :numCasas, :numHoteles , :titulo

    #  def initialize (num,c,type,prop)
    #   @numeroCasilla=num
    #  @coste=c
    # @numCasas=0
    #@numHoteles=0
    #@tipo=type
    #@titulo=prop
    #end
    def initialize(num,c,type)
      @numeroCasilla=num
      @coste=c
      @tipo=type
    end
    
    def soyEdificable
      #soy_edificable = false
      #if (@tipo == TipoCasilla::CALLE)
      #  soy_edificable = true
      #end
      #return soy_edificable
      return false
    end
    
    def tengoPropietario
      return false
    end
    
    def to_s
      cost=""
      if @coste != 0
        cost="Coste: #{@coste}."
      end
      return "Numero de casilla: #{@numeroCasilla}.#{cost}\nTipo de casilla: #{@tipo}\n\n"
    end
    
    attr_reader :coste,:numeroCasilla,:tipo
    # def self.creaCalle(num,c)
    #  new(num,c,TipoCasilla::CALLE,nil)
    #end

    # def self.creaCasilla(num,c,type)
    #  new(num,c,type,nil)
    #end


    #   def to_s
    # if @tipo == TipoCasilla::CALLE
    #      pro= "Título de propiedad :\n#{@titulo}"
    #    else
    #      pro=""
    #    end
    #    "Numero de casilla: #{@numeroCasilla}.\nCoste de compra: #{@coste}\n" +
    #     "Número de casas construidas: #{@numCasas}.\nNúmero de hoteles construidos: #{@numHoteles}.\n" +
    #     "Tipo de casilla: #{@tipo}.\n#{pro}\n"
    # end

    #   def calcularValorHipoteca
    #    hipotecaBase=@titulo.hipotecaBase
    #    cantidadRecibida = hipotecaBase + Integer( @numCasas * 0.5 * hipotecaBase + @numHoteles * hipotecaBase)
    #    return cantidadRecibida
    #  end

    #  def cancelarHipoteca
    #    @titulo.hipotecada=false
    #    aPagar=getCosteHipoteca
    #    return aPagar
    #  end

    #def cobrarAlquiler
    #  costeAlquilerBase=@titulo.alquilerBase
    #   costeAlquiler = costeAlquilerBase + @numCasas * 0.5 + @numHoteles * 2
    #   @titulo.cobrarAlquiler(costeAlquiler)
    #   return costeAlquiler
    # end

    # def edificarCasa
    #   @numCasas+=1
    #   costeEdificarCasa=getPrecioEdificar
    #   return costeEdificarCasa
    # end

    # def edificarHotel
    #   @numHoteles+=1
    #   costeEdificarHotel=getPrecioEdificar
    #   return costeEdificarHotel
    #  end

    #def estaHipotecada
    #  return @titulo.hipotecada
    # end

    #    def getCosteHipoteca
    #      coste = calcularValorHipoteca * 1.10
    #     return Integer(coste)
    #   end

    #   def getPrecioEdificar
    #     return @titulo.precioEdificar

    #   end

    #   def hipotecar
    #     @titulo.hipotecada=true
    #     cantidadRecibida=calcularValorHipoteca
    ##     return cantidadRecibida
    #   end

    #   def propietarioEncarcelado
    #     return @encarcelado
    #   end

    #   def sePuedeEdificarCasa(factorEspeculador)
    #     return @numCasas < 4*factorEspeculador
    #   end

    #   def sePuedeEdificarHotel(factorEspeculador)
    #     return @numHoteles < 4*factorEspeculador && @numCasas == 4*factorEspeculador
    #   end

 

    #   def tengoPropietario
    #     if @titulo != nil
    #       return @titulo.propietario != nil
    #     end
    #   end

    #   def venderTitulo
    ##     precioCompra = @coste + (@numCasas + @numHoteles)*@titulo.precioEdificar
    #     precioVenta = precioCompra + @titulo.factorRevalorizacion * precioCompra
    #     @titulo.propietario= nil
    #     @numHoteles=0
    #     @numCasas=0
    #     return Integer(precioVenta)
    #   end
  end
end
